advanced/demos/python_GUI.py

```diff
@@ -0,0 +1,318 @@
+        self.camera = [0, 25, 3]
+
+    #determines the 3D center of a face (used for determining draw order)
+    def calculate_face_center(self, mesh, i):
+        totals = [0, 0, 0]
+        num = 0
+        for p in mesh.faces[i]:
+            totals[0] += mesh.vertices[p][0]
+            totals[1] += mesh.vertices[p][1]
+            totals[2] += mesh.vertices[p][2]
+            num += 1
+        return [t / num for t in totals]
+
+    #Determined the color of the face by calculating the projection of the face vector and the incident vector (vector
+    #from light source to face center
+    def generate_shading(self, face_vector, face_pos, light_src, base_color):
+        face_vector = numpy.array(face_vector)
+        face_pos = numpy.array(face_pos)
+        light_src = numpy.array(light_src)
+        incident_vector = light_src - face_pos
+        proj_inc_face = (numpy.dot(face_vector, incident_vector)) * incident_vector / (
+                numpy.linalg.norm(incident_vector) ** 2)
+        color = [str(hex(int(base_color[i] * numpy.linalg.norm(proj_inc_face)))).replace('0x', '') for i in range(3)]
+        for i in range(3):
+            if len(color[i]) < 2:
+                color[i] = "0" + color[i]
+        return "#" + "".join(color)
+
+    #draws mesh as a wireframe by applting a perspective transform on the 3d point to translate it to 2d
+    #also applies transformation for the camera rotation and position
+    #offset is the offset of the image, not the object (needed so it is not drawn at (0,0))
+    def draw_mesh_wireframe(self, mesh, colors, horizon, offsets, camera, camera_angle):
+        i = 0
+        for m in mesh:
+            for e in m.edges:
+                v1 = [x - y for x, y in zip(m.vertices[e[0]], camera)]
+                TRot = [[1, 0, 0], [0, numpy.cos(camera_angle), -1 * numpy.sin(camera_angle)],
+                        [0, numpy.sin(camera_angle), numpy.cos(camera_angle)]]
+                v1 = numpy.dot(TRot, v1)
+                transform1 = [[-1 * horizon / v1[2], 0, 0], [0, -1 * horizon / v1[2], 0]]
+                p1 = numpy.dot(transform1, v1)
+                p1[0] += offsets[0]
+                p1[1] += offsets[1]
+                v2 = [x - y for x, y in zip(m.vertices[e[1]], camera)]
+                v2 = numpy.dot(TRot, v2)
+                transform2 = [[-1 * horizon / v2[2], 0, 0], [0, -1 * horizon / v2[2], 0]]
+                p2 = numpy.dot(transform2, v2)
+                p2[0] += offsets[0]
+                p2[1] += offsets[1]
+                self.canvas.create_line(p1[0], p1[1], p2[0], p2[1], fill=colors[i % len(colors)])
+            i += 1
+
+    #draws mesh as a solid using a similar technique, however it differs in that it calculates the position of each face
+    #first, sorts them by distance from camera and draws in that order (this is so objects appear at the correct depth
+    def draw_mesh_solid(self, mesh, mesh_colors, horizon, offsets, camera, camera_angle, light):
```

Comment:
The canvas/etc APIs don't have an option to enable depth buffering and depth testing? I guess I'm just used to seeing most drawing libraries sit on top of OpenGL. If not, then this is (sadly) the only way to do it...